### ISSUE TITLE
[Optimization RFC] Safe typed slice reinterpretations (as_f32_slice) for SIMD vectorization

### ISSUE THREAD

**Issue by sanohiro** [sanohiro/align#1064]
### Problem

Scanning 151,936 vocabulary logits in `greedy` decode and `select` takes ~111 µs per token because the loop runs 100% scalar on both x86-64 and ARM64.

`alignc explain-opt` reports that LLVM LoopVectorizer refuses vectorization due to:
1. Logits held in `buffer.bytes()` as `slice<u8>` and accessed via scalar `f32_le(offset)` with dual induction variables (`offset = offset + 4`, `i = i + 1`), preventing typed SIMD vector loads.
2. In-loop early exit NaN/Inf checks (`if (bits & 0x7f800000) == 0x7f800000 { return Err(Error.Invalid) }`), triggering `Auto-vectorization of loops with potentially faulting load is not supported`.
3. Paired argmax state updates `(best_val, best_idx)` in scalar control flow, triggering `a value carried between iterations wasn't recognized as a reduction`.

### Architectural Decision: Option A (Safe Typed Slice Views) over Option B (Ad-hoc Primitives)

**Option B (adding `slice.argmax()` as an ad-hoc intrinsic)** is rejected as a piecemeal workaround:
- Adding specialized algorithms like `argmax` to the compiler/stdlib sets a bad precedent. Next time someone needs `argmin`, `top_2`, `dot`, `variance`, or `norm`, another ad-hoc intrinsic would have to be added.
- It violates Align's core load-bearing invariant:
  > **Data-oriented core:** normal array and slice pipelines must lower well to SIMD, cache-friendly execution, and future accelerators.

**Option A (Safe Typed Slice Views)** addresses the fundamental root cause:
- Provide safe checked reinterpretation from byte slices to typed numeric slices:
  ```align
  slice<u8>.as_f32_slice() -> Result<slice<f32>, Error>
  slice<u8>.as_i32_slice() -> Result<slice<i32>, Error>
  slice<u8>.as_f64_slice() -> Result<slice<f64>, Error>
  slice<u8>.as_i64_slice() -> Result<slice<i64>, Error>
  ```
- Reinterpretation validates:
  1. Pointer alignment is an exact multiple of the element type's alignment (e.g. 4 bytes for `f32`).
  2. Byte length is an exact multiple of the element size (`len % 4 == 0`).
- Lifetime & Ownership: The resulting typed slice borrows the same backing region as the source byte slice (zero copy, zero allocation).

### Impact & Vectorization Evidence

In `align-llm` benchmarking, once logits are exposed as an aligned `slice<f32>`:
- A simple clean loop or reduction over `slice<f32>` allows LLVM to auto-vectorize into 4x unrolled 256-bit AVX2 SIMD (`vmaxps`, 32 floats per iteration) on x86-64 and NEON (`fmaxnm`, 16 floats per iteration) on Apple Silicon ARM64.
- Latency drops from 111 µs to <= 15–20 µs (5–7x speedup) without adding specialized compiler intrinsics for argmax.

### Acceptance Criteria

- `slice<u8>.as_f32_slice()` (and sibling numeric types) type-checks and validates pointer alignment and byte length.
- An unaligned pointer or non-multiple length produces a clear `Result::Err(Error::Invalid)`.
- Normal loops over the resulting `slice<f32>` auto-vectorize cleanly in LLVM LoopVectorizer.
- Verified with `align-llm/scripts/bench-runtime-greedy`.

**Comment by sanohiro** [comment]
### Architectural Rationale & Why This Avoids Ad-Hoc Compiler Growth

#### 1. Why Option B (`slice.argmax()`) was ad-hoc (場当たり的)
When LLVM failed to auto-vectorize greedy token selection in `align-llm`, the immediate workaround was to propose adding a specialized `argmax()` primitive to the compiler and runtime.

However:
- Adding specialized algorithms as compiler/runtime builtins is an ad-hoc band-aid. If we add `argmax`, what happens when a program needs `argmin`, `top_2`, `variance`, `dot`, `l2_norm`, or `cumsum`? Adding a new compiler intrinsic for every domain-specific algorithm turns the compiler into a bloated kitchen sink.
- It violates Align's core load-bearing invariant:
  > **Data-oriented core:** normal array and slice pipelines must lower well to SIMD, cache-friendly execution, and future accelerators.

#### 2. Why Option A (`slice<u8>.as_f32_slice()`) is the necessary architectural solution
- **Targeting the True Root Cause**: The barrier to auto-vectorization was NOT the algorithm itself, but the data representation. Logits were trapped in a `slice<u8>` byte slice, forcing byte-by-byte offset arithmetic (`f32_le(offset)`) and dual induction variables that obscured memory contiguousness from LLVM's LoopVectorizer.
- **Enabling Natural SIMD on Normal Align Code**: By providing safe, zero-copy, alignment-verified typed slice views (`slice<u8>.as_f32_slice() -> Result<slice<f32>, Error>`), standard Align loops and reductions over `slice<f32>` naturally compile into 4x unrolled 256-bit AVX2 SIMD (`vmaxps`, 32 floats/iter) and ARM64 NEON (`fmaxnm`, 16 floats/iter).
- This empowers developers to write performant numerical algorithms directly in ordinary Align code rather than requiring compiler modifications for every new workload.

**Comment by sanohiro** [comment]
Review of this issue as one part of a SIMD-by-default policy for Align (evidence: the align-llm binary audit, issues #1069–#1087, plus fact-finding experiments on Align `8c8bfbc7` / LLVM 22.1.8, Apple M1). The yardstick is the best form for Align as a general language; align-llm is evidence only. Measurements referenced below are posted in the following comment.

## 2. #1064 — safe typed slice views

Verdict: **right direction (no per-algorithm intrinsics), but the shape re-proposes what plan 62
rejected without answering it, the API is still enumerated per type, and the acceptance criterion
is measured false.**

1. **Reconcile with plan 62 explicitly.** `docs/impl/62-decode-optimization-plan.md:45` rejected
   `as_f32_le()` for three reasons. Each has an answer, and the issue should record them:
   - *alignment* → checked at runtime, `Option<slice<T>>`/`Result`; a mis-aligned sub-slice returns
     none rather than aborting. `buffer`/`array` payloads come from `malloc` (16-aligned) so views at
     offset 0 always succeed; sub-slices are checked.
   - *byte order* → a view is a **native-order** reinterpretation by definition. Align hard-codes
     little-endian hosts (`lib.rs:14041`) but the spec's rule is that every multi-byte access names its
     order. Keep the rule: name the order in the view (`_le`), and make the compiler reject a view
     whose named order is not the target's native order ("cannot view big-endian data on a
     little-endian target; decode instead"). The wire format stays explicit; zero-copy stays
     guaranteed; nothing is hidden.
   - *access authority* → for a view over an Align-owned borrow, authority is the borrow's:
     `slice<u8>` gives a read-only `slice<T>`, `mut slice<u8>` a writable one. Plan 62's concern was
     raw/FFI memory, which stays with `resource.view_from_raw` (plan 61). Two routes, one rule each:
     Align-owned bytes → checked view; foreign memory → owner-published typed view.
2. **One generic operation, not four methods.** `as_f32_slice/as_i32_slice/as_f64_slice/as_i64_slice`
   is the same per-width enumeration the `<ty>_le/_be` accessor family already has (36 methods).
   Define one `bytes.view_le<T>() -> Option<slice<T>>` for the closed scalar set
   (`u16..u64, i16..i64, f32, f64`) plus the writable form and the inverse `slice<T>.as_bytes()`.
   The existing `<ty>_le(off)` accessors then become "view then index" in the spec, which is also
   the path that removes B1/B2's dual induction variables.
3. **Fix the acceptance criterion.** "Clean loops over the resulting `slice<f32>` auto-vectorize" is
   false today (190.4 µs, identical to bytes) and does not become true through this issue. The
   view is a *prerequisite* for reaching `slice<T>.load(i) -> vecN<T>` and for typed pipelines; the
   vectorization itself is owned by B1/B2/B4 (#1079, #1081, #1082). Acceptance should be:
   zero-copy view construction; `vecN` load/store reachable from `buffer.bytes()`; the greedy-argmax
   target moves to the umbrella below.
4. **Early exits (B11).** #1064's NaN/Inf early-return is an argmax-specific workaround. The language
   answer is a canonical form the compiler can split: an exit test that does not feed the reduction
   is hoisted into its own vectorizable pass (or-reduction of the exponent mask) ahead of the main
   loop. That is a MIR transformation, not a library function.


The umbrella contract that ties this to the other blockers is filed as #1088 ("Vectorization contract").

**Comment by sanohiro** [comment]
Measurements from the align-llm binary-optimization audit that bear directly on this issue's acceptance
criteria. Everything below was produced with the pinned toolchain — Align
`8c8bfbc7a3169e84ecc8415f5149ab8c61afe863` (alignc 0.7.5, LLVM 22.1.8), Apple M1, `--profile release`,
default `--target-cpu baseline` — on 152,064 `f32` (300 calls, best of 3). No new API is proposed here;
Option A and the rejection of Option B both hold up under measurement, and this is just evidence about
what else has to be true for the last acceptance criterion to pass.

### 1. Option A removes the route cost, which is real and large

The cost of *getting* from ggml's `slice<u8>` logits to typed `f32` today, written as an explicit
conversion loop (`b.f32_le(i*4)` into an `out slice<f32>`):

| form | µs/call | GB/s |
|---|---|---|
| `slice<u8>` → `out slice<f32>`, as the compiler emits it | **59.4** | 20.5 |
| same loop with the in-loop bounds checks hoisted **and** `noalias` on the borrow header | **12.6** | 96.5 |

As emitted, that loop is 10 instructions per single `f32`, with the view header reloaded twice per element
and one surviving range check:

```
2c: ldr  x2, [x0, #0x8]      ; LEN reloaded
30: add  x9, x8, #0x4
34: cmp  x9, x2
38: b.gt <range trap>
3c: ldr  x11, [x0]           ; PTR reloaded
40: subs x10, x10, #0x1
44: ldr  s0, [x11, x8]
48: str  s0, [x1, x8]
4c: mov  x8, x9
50: b.ne 0x2c
```

A zero-copy checked view makes this number **zero**, which is a good independent argument for Option A
over any conversion helper.

### 2. But a typed view alone does not make the argmax loop vectorize

Same host, same run, three argmax/max variants:

| form | µs/call | GB/s |
|---|---|---|
| argmax over `slice<u8>` + `f32_le` (align-llm's shape today) | 190.6 | 3.19 |
| argmax over an **already-typed** `slice<f32>` (`if xs[i] > best`) | **190.4** | 3.20 |
| `slice<f32>.max()` pipeline terminal | 190.7 | 3.19 |

Handing the loop a `slice<f32>` changes the time by **0.1 %**. The remaining blockers are not the data
representation:

- the borrowed view header is re-materialized at every use and a `borrow` header parameter carries no
  `noalias`, so a loop that also stores is assumed to invalidate it;
- two or more surviving in-loop bounds checks switch LoopVectorize off entirely (threshold measured at
  exactly 1).

Both are conjunctive, measured on the conversion kernel above, four variants through the identical
`default<O2>`:

| variant | `vector.body` | `<4 x float>` ops | µs/call |
|---|---|---|---|
| as emitted | 0 | 0 | 59.4 |
| in-loop checks hoisted only | **0** | 0 | 59.4 |
| `noalias` on the borrow header only | **0** | 0 | 59.4 |
| both | **2** | **10** | **12.6** |

So this issue's last acceptance criterion — *"normal loops over the resulting `slice<f32>`
auto-vectorize cleanly in LLVM LoopVectorizer"* — depends on two separately filed compiler issues:
*"[Codegen] Materialize a borrowed slice/array view header once and state its alias facts"*
(#TBD-13) and *"[MIR] Fuse the bounds check into one unsigned compare and eliminate
monotone-induction in-loop checks"* (#TBD-16). Neither of them alone is measurable; the pair is 4.7×
on that kernel. Worth landing the typed view and those two together, or the criterion will not be
demonstrable even though the view is correct.

### 3. On the `fmaxnm` / `vmaxps` expectation specifically

An `if v > best` comparison (and the `slice<f32>.max()` pipeline terminal) lowers to

```llvm
  %fgt = fcmp ogt float %slcload, %_1.035
  %_1.1 = select i1 %fgt, float %slcload, float %_1.035
```

An `fcmp` + `select` pair carries no NaN semantics, so LLVM will not form a floating-point max reduction
from it — that is why the measured 190.4 µs above is at parity with a clang `-O2` strict-IEEE C twin of
the same reduction (~190.5 µs). No typed view changes this.

The encouraging part: Align's **scalar** `a.max(b)` already lowers to `llvm.maximum`, and LLVM 22 *does*
widen that. Writing the same reduction as `best = best.max(xs[i])` over a `slice<f32>` gives, at
`--target-cpu baseline`, with no fast-math:

```llvm
  %6 = tail call <4 x float> @llvm.maximum.v4f32(<4 x float> %vec.phi, <4 x float> %wide.load)
  %7 = tail call <4 x float> @llvm.maximum.v4f32(<4 x float> %vec.phi27, <4 x float> %wide.load28)
```

```
0000000000000054	fmax.4s	v0, v0, v2
0000000000000058	fmax.4s	v1, v1, v3
0000000000000060	fmax.4s	v0, v0, v1
0000000000000068	fmaxv.4s	s0, v0
```

and measures **12.0 µs/call** — **16×** the `fcmp`+`select` form, with identical NaN and ±0
behaviour. So the vectorized max reduction is already legal and already reachable; it is the
*pipeline terminal* and the comparison form that do not use it. That inconsistency is filed as
*"[Language RFC] One floating-point reduction semantics: uniform min/max lowering plus scoped
reassociation"* (#TBD-15).

One small correction to the criterion's wording: Align's chosen instruction is `fmax` / `llvm.maximum`
(NaN-propagating IEEE 754-2019), deliberately not `fmaxnm` / `llvm.maxnum` — see the rationale comment at
`crates/align_codegen_llvm/src/lib.rs:11463-11466`. Expecting `fmax.4s` + `fmaxv.4s` on aarch64 (and
`maxps`-family with the NaN-correct blend on x86-64) rather than `fmaxnm`/`vmaxps` will make the criterion
checkable.

### 4. For argmax specifically (index-carrying), one front-end restriction also applies

A hand-written vectorized argmax over `slice<f32>` measures **47.5 µs/call** (4.0× over the 190.4 µs
scalar form) with `fcmgt.4s` + two `bit.16b` per 4 lanes — but only if the index vector is kept in `f32`,
because `select` currently requires the mask's element type to equal the blended vectors':

```
error: 'select' mask mask4<f32> does not match the vectors vec4<i32>
```

The `f32`-index workaround is exact only below 2^24. Filed as *"[Sema] Make a vector mask
structural: blend vectors of matching lane count and element width"* (#TBD-18) — a one-arm
relaxation in sema that codegen already ignores. That is what lets ordinary Align code express a
vectorized argmax, which is this issue's stated goal, without any `argmax` intrinsic.

### Summary of the dependency, for sequencing

| step | measured effect on a 152,064-logit selection |
|---|---|
| typed view (this issue) | removes the 59.4 µs route cost; argmax itself unchanged at 190.4 µs |
| + uniform min/max lowering | max reduction 190.7 → 12.0 µs |
| + view-header/alias facts **and** bounds-check shape | conversion-style loops 59.4 → 12.6 µs; makes "clean loops vectorize" true in general |
| + structural mask rule | hand-written vec4 argmax 190.4 → 47.5 µs, exact for all lengths |

### One open question on the proposed spelling: byte order

`docs/impl/62-decode-optimization-plan.md:45` rejected a naive `as_f32_le()` for *"pretending all byte
storage has native alignment, byte order and access authority"*. The revised surface answers the
alignment and length halves explicitly (pointer alignment and `len % 4 == 0`, validated at run time),
but `as_f32_slice()` names **neither** the byte order nor the host-native assumption, while
`docs/language-spec.md:727-735` makes every other multi-byte access endian-explicit (`h.u32_le(0)`,
`out.put_u64_be(n)`, suffix-free only for `u8`/`i8`). A reinterpreted view differs from those reads in
kind — it cannot swap bytes without copying, so it is a native-order view by construction rather than
by choice of suffix.

So the question for the design, not a proposal: should the name carry that fact (a `_ne`-style
suffix, or a documented statement that a reinterpreting view is native-order-only and therefore the
one exception to the `_le`/`_be` rule)? Whatever the answer, none of the measurements above
depend on it: they are about what the *loop* over the resulting typed slice compiles to.

**Comment by sanohiro** [comment]
Implemented and merged.

- Design/contract: #1155
- Compiler implementation: #1156 (merge `dfcfd11f0b076320f3f9bfbf1f837a4130092964`)
- Shipped surface: checked zero-copy `bytes.view_le<T>()` views over the closed scalar domain, writable authority preservation, and allocation-free `slice<T>.as_bytes()` inverse views
- Safety/portability: alignment and byte-length checks, multiplication-overflow trap, provenance/lifetime preservation, checked-HIR/MIR validation, and little-endian admission before every cache or LLVM entry point
- Optimization contract: explicit vector loads/stores are reachable through the view, but arbitrary consumer-loop vectorization is not promised; final SIMD shape remains visible through optimization diagnostics and object inspection

Required Linux x86_64, Linux ARM64, and macOS Apple Silicon CI passed. align-llm adoption and its named benchmarks remain consumer-owned and are recorded as pending in Request 93.
